### DISCUSSION TITLE
Is it possible to have multiple cells in dragStartClone (stencil)?

### DISCUSSION THREAD

**Discussion by frnora** [clientIO/joint#2126] [Q&A] [answered]
### Introduction

Hi,

I would like to have a more complex "object" (subgraph?) as the `dragStartClone` of my stencil.

The stencil shape still has to remain a simple element but dragging it should create basically a tree-branch structure: multiple elements connected by links. I can create the structure without problems, of course, after placing it on the paper but I would also like to show the branch while dragging so the user can see where it is going to be placed and how it will look like.

Seems like the `dragStartClone` always has to return an element. Any way to show multiple elements/links? Do I need to render it to some image and then show that one? Something similar to what you are doing in `addStencilShapeWithPreview` in the Dynamic Stencil example?

Thank you!

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS+

**Comment by kumilingus** [comment, marked as answer]
At the moment, using the image approach from the [Dynamic Stencil](https://www.jointjs.com/demos/dynamic-stencil) example is the safest.
It works with all other JointJS features (e.g. embedding, snaplines).

---

Here's a workaround. It adds a transparent rectangle on `dragStartClone` (to determine the size of the preview).
And then on `element:dragstart`, when it has access to the preview graph, it adds the desired cells into it.
Later on `element:dragend` it cancels the default drop and adds the cells into the main graph.

```js
stencil.options.dragStartClone = (cell) => {
    const background = new joint.shapes.standard.Rectangle({
        size: { width: 100, height: 100 },
        attrs: {
            body: {
                stroke: 'none',
                fill: 'transparent'
            }
        }
    });
    return background;
};

stencil.on('element:dragstart', (cloneView) => {
    const previewGraph = cloneView.paper.model;
    const circle1 = new joint.shapes.standard.Circle({
        position: { x: 10, y: 10 },
        size: { width: 20, height: 20 },
        attrs: {
            body: {
                fill: 'red'
            }
        }
    });
    const circle2 = new joint.shapes.standard.Circle({
        position: { x: 70, y: 70 },
        size: { width: 20, height: 20 },
        attrs: {
            body: {
                fill: 'blue'
            }
        }
    });
    previewGraph.addCells([circle1, circle2]);
});

stencil.on('element:dragend', (cloneView, evt, cloneArea, validDropTarget) => {
    if (!validDropTarget) return;
    stencil.cancelDrag({ dropAnimation: false });
    const { x, y } = cloneArea;
    const circle1 = new joint.shapes.standard.Circle({
        position: { x: x + 10, y: y + 10 },
        size: { width: 20, height: 20 },
        attrs: {
            body: {
                fill: 'red'
            }
        }
    });
    const circle2 = new joint.shapes.standard.Circle({
        position: { x: x + 70, y: y + 70 },
        size: { width: 20, height: 20 },
        attrs: {
            body: {
                fill: 'blue'
            }
        }
    });
    mainGraph.addCells([circle1, circle2]);
});
```

**Reply by frnora** [reply]
Good idea, thank you, Roman!
